File: 1. 단계별로_풀어보기/28. 위상 정렬/1. 2252.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [N, M] = inputArr[0].trim().split(' ').map(Number)
const seqArr = inputArr.slice(1).map(row => row.trim().split(' ').map(Number))

// remainArr[i] : i 보다 앞에 있는 사람중 남은 사람의 수
const remainArr = Array.from({length: N + 1}).fill(0)
const mapArr = Array.from({length: N + 1}, () => {
  return {}
})
seqArr.forEach(seq => {
  const [A, B] = seq
  mapArr[A][B] = 1
  remainArr[B]++
})

// nextArr[i]: i 보다 뒤에 있는 사람의 목록
const nextArr = mapArr.map((mapInfo, mapIdx) => Object.keys(mapInfo))

// 방문큐. 배열로 선언한다.
const visitQueue = Array.from({length: 32001}).fill(0)
let nowIdx = 0
let insertIdx = 0

// 본인보다 확실하게 앞에 있는 사람은 없는 사람들을 방문큐에 넣는다.
for (let people = 1; people <= N; people++) {
  if (remainArr[people] === 0) {
    visitQueue[insertIdx++] = people
  }
}

// 출력 변수
let result = ''

// 위상정렬을 수행한다.
while (nowIdx < insertIdx) {
  const now = visitQueue[nowIdx++]
  result += now + ' '

  // now 보다 뒤에있는게 확실한 사람들의 remainArr 값을 1 줄인다.
  nextArr[now].forEach(next => {
    remainArr[next]--

    // 만약, next 보다 확실하게 큰 사람이 0명이 된 경우
    // 방문큐에 넣는다.
    if (remainArr[next] === 0) {
      visitQueue[insertIdx++] = next
    }
  })
}

// 출력
process.stdout.write(result)

File: 1. 단계별로_풀어보기/20~29/28. 위상 정렬/1. 2252.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

N, M = map(int, input().split())

# nextArr[i]: i번째 사람보다 확실히 작은 사람들의 목록
# push 연산과 검색만 하기 때문에 배열로 선언해도 시간초과 안난다.
nextArr = list([] for i in range(N + 1))

# remainArr[i]: i번째 사람보다 확실히 큰 사람들의 수
remainArr = list(0 for i in range(N + 1))

for _ in range(M):
  A, B = map(int, input().split())
  nextArr[A].append(B)
  remainArr[B] += 1

# 방문큐
visitQueue = deque()

# 본인보다 확실히 큰 사람이 없는 사람들을 방문큐에 넣는다.
for people in range(1, N + 1):
  if remainArr[people] == 0:
    visitQueue.append(people)

while visitQueue:
  now = visitQueue.popleft()
  output(f"{now} ")

  for next in nextArr[now]:
    remainArr[next] -= 1
    
    # 더 이상 next 보다 확실히 큰 사람은 없으면 방문큐에 넣는다
    if remainArr[next] == 0:
      visitQueue.append(next)